File: leetcode/917. Reverse Only Letters/917.cpp
// https://helloacm.com/how-to-reverse-only-letters-in-a-string/
// https://leetcode.com/problems/reverse-only-letters/

class Solution {
public:
    string reverseOnlyLetters(string S) {
        auto len = S.size();
        int i = 0, j = len - 1;
        while (i < j) {
            while ((i < j) && (!isalpha(S[i]))) i ++;
            while ((i < j) && (!isalpha(S[j]))) j --;
            swap(S[i ++], S[j --]);
        }
        return S;
    }
};

File: leetcode/917. Reverse Only Letters/917.java
// https://helloacm.com/how-to-reverse-only-letters-in-a-string/
// https://leetcode.com/problems/reverse-only-letters/

class Solution {
    public String reverseOnlyLetters(String S) {
        int i = 0, j = S.length() - 1;
        char[] r = S.toCharArray();
        while (i < j) {
        	while (i < j && !Character.isLetter(S.charAt(i))) ++i;
        	while (i < j && !Character.isLetter(S.charAt(j))) --j;
        	char t = r[i];
        	r[i] = r[j];
        	r[j] = t;
        	i ++;
        	j --;
        }
        return String.valueOf(r);
    }
}
